lib/logscheduler.c

```diff
@@ -0,0 +1,429 @@
+    {
+      if (iv_list_empty(&thread_state->batch_by_partition[partition_index]))
+        continue;
+
+      /* form the new batch, hand over the accumulated elements in batch_by_partition */
+      LogSchedulerBatch *batch = _batch_new(&thread_state->batch_by_partition[partition_index]);
+      INIT_IV_LIST_HEAD(&thread_state->batch_by_partition[partition_index]);
+
+      /* add the new batch to the target partition */
+
+      LogSchedulerPartition *partition = &self->partitions[partition_index];
+
+      _partition_add_batch(partition, batch);
+
+
+    }
+  thread_state->num_messages = 0;
+  return NULL;
+}
+
+static void
+_queue_thread(LogScheduler *self, LogSchedulerThreadState *thread_state, LogMessage *msg,
+              const LogPathOptions *path_options)
+{
+  if (thread_state->num_messages == 0)
+    main_loop_worker_register_batch_callback(&thread_state->batch_callback);
+
+  gint partition_index = _get_partition_index(self, thread_state, msg);
+
+  LogMessageQueueNode *node;
+  node = log_msg_alloc_queue_node(msg, path_options);
+  iv_list_add_tail(&node->list, &thread_state->batch_by_partition[partition_index]);
+  thread_state->num_messages++;
+  log_msg_unref(msg);
+}
+
+
+static void
+_thread_state_init(LogScheduler *self, LogSchedulerThreadState *state)
+{
+  worker_batch_callback_init(&state->batch_callback);
+  state->batch_callback.func = _flush_batch;
+  state->batch_callback.user_data = self;
+
+  for (gint i = 0; i < self->options->num_partitions; i++)
+    INIT_IV_LIST_HEAD(&state->batch_by_partition[i]);
+}
+
+static void
+_init_thread_states(LogScheduler *self)
+{
+  for (gint i = 0; i < self->num_threads; i++)
+    {
+      _thread_state_init(self, &self->thread_states[i]);
+    }
+}
+
+static void
+_init_partitions(LogScheduler *self)
+{
+  for (gint i = 0; i < self->options->num_partitions; i++)
+    {
+      _partition_init(&self->partitions[i], self->front_pipe);
+    }
+}
+
+static void
+_free_partitions(LogScheduler *self)
+{
+  for (gint i = 0; i < self->options->num_partitions; i++)
+    {
+      _partition_clear(&self->partitions[i]);
+    }
+}
+
+gboolean
+log_scheduler_init(LogScheduler *self)
+{
+  return TRUE;
+}
+
+void
+log_scheduler_deinit(LogScheduler *self)
+{
+}
+
+void
+log_scheduler_push(LogScheduler *self, LogMessage *msg, const LogPathOptions *path_options)
+{
+  gint thread_index = main_loop_worker_get_thread_index();
+
+  if (!self->front_pipe ||
+      self->options->num_partitions == 0 ||
+      thread_index < 0 ||
+      thread_index >= self->num_threads)
+    {
+      _reinject_message(self->front_pipe, msg, path_options);
+      return;
+    }
+
+  LogSchedulerThreadState *thread_state = &self->thread_states[thread_index];
+  _queue_thread(self, thread_state, msg, path_options);
+}
+
+LogScheduler *
+log_scheduler_new(LogSchedulerOptions *options, LogPipe *front_pipe)
+{
+  gint max_threads = main_loop_worker_get_max_number_of_threads();
```

Comment:
This includes destination threads as well.

Shouldn't we allocate thread state objects only for input threads?

Reply:
I am starting to think that we should, yeah. I got rid off using separate ID space for destinations not long ago.

Our per-thread state is 48 bytes at the moment, so with 64 threads in total, it's 3k of memory per LogScheduler.

I don't think this is a pressing issue for now. We also have the InputQueue for LogQueueFifo, which is 64 bytes and is also a per-thread struct.

If we were to run on 256+ threads we might want to do something  with this.